docs/design.md

```diff
@@ -0,0 +1,279 @@
+# Indexer Reference Provider Design
+The goal of this reference implementation is to flesh out all the protocol details
+for data providers to be able to interact with indexer nodes. 
+
+## Overview
+- Data providers keep a local index of all the data they are providing.
+- Data providers are responsible for the advertisement of updates related to the data they are providing.
+        - Data providers must be able to uniquely identify the batches of data they are providing, and/or the
+        announcements they broadcast to the network in order to be able to serve indexer-node data ingestion requests.
+- Indexer nodes listen to advertisement from data providers. When they come across a previously unseeend
+advertisement for a data provider, they trigger a new ingestion session to sync with the provider. 
+
+## Data Structures
+Data providers and indexer nodes can use any internal representation they want for indexed data. They only
+data structures enforced by the protocol are in the interaction between the two.
+
+### Advertisements
+Advertisements are used by data providers to share updates about the data they are providing. Ideally,
+these advertisements are published in a shared pub-sub channel between providers and indexer nodes. However,
+some indexer nodes may choose to provide a dedicated endpoint to allow certain authenticated providers to push announcements and updates directly to them.
+- `Pull` Advertisements: Indexer-nodes pull advertisement from a shared pub-sub channel.
```

Comment:
Need to add the method where we read the advertisements from an S3 bucket, or a file, IPFS, or whatever.  I think we can call it the `Load` method?  Unless you think of something better.  Not sure if that goes in this section of the doc or in the "Advertisement Protocol" section, or both.

Reply:
Good catch, I was missing this. I added the new functions to the interface. Let me know what you think.